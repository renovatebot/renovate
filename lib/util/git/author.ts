import addrs from 'email-addresses';
import { logger } from '../../logger';
import { regEx } from '../regex';
import type { GitAuthor } from './types';

export function parseGitAuthor(input: string): GitAuthor | null {
  let result: GitAuthor | null = null;
  if (!input) {
    return null;
  }
  try {
    result = addrs.parseOneAddress(input);
    if (result) {
      return result;
    }
    let massagedInput: string | undefined;
    let massagedBotEmail = false;
    if (input.includes('<') && input.includes('>')) {
      // try wrapping the name part in quotations
      massagedInput = '"' + input.replace(regEx(/(\s?<)/), '"$1');
    }
    if (input.includes('[bot]@')) {
      // invalid github app/bot addresses
      massagedInput = (massagedInput ?? input).replace('[bot]@', '@');
      massagedBotEmail = true;
    }
    if (!massagedInput) {
      return null;
    }
    const parsed = addrs.parseOneAddress(massagedInput) as addrs.ParsedMailbox;
    if (parsed?.address) {
      result = {
        name: parsed.name ?? input.replace(regEx(/@.*/), ''),
        address: parsed.address,
      };
      if (massagedBotEmail) {
        result.address = result.address?.replace('@', '[bot]@');
      }
      return result;
    }
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ err }, 'Unknown error parsing gitAuthor');
  }
  // give up
  return null;
}
