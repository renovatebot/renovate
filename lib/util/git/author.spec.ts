import { parseGitAuthor } from './author';

describe('util/git/author', () => {
  describe('parseGitAuthor', () => {
    it('returns null if empty email given', () => {
      expect(parseGitAuthor(undefined as never)).toBeNull();
    });

    it('handles a normal address', () => {
      expect(parseGitAuthor('renovate@whitesourcesoftware.com')).not.toBeNull();
    });

    it('parses bot email', () => {
      expect(parseGitAuthor('renovate[bot]@users.noreply.github.com')).toEqual({
        address: 'renovate[bot]@users.noreply.github.com',
        name: 'renovate[bot]',
      });
    });

    it('parses bot name and email', () => {
      expect(
        parseGitAuthor('renovate[bot] <renovate[bot]@users.noreply.github.com>')
      ).toEqual({
        address: 'renovate[bot]@users.noreply.github.com',
        name: 'renovate[bot]',
      });
    });

    it('escapes names', () => {
      expect(parseGitAuthor('name [what] <name@what.com>')?.name).toBe(
        `name [what]`
      );
    });

    it('tries again and fails', () => {
      expect(parseGitAuthor('foo<foo>')).toBeNull();
    });

    it('gives up', () => {
      expect(parseGitAuthor('a.b.c')).toBeNull();
    });
  });
});
