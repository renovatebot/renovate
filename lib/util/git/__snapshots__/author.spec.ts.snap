// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/git/author parseGitAuthor escapes names 1`] = `"name [what]"`;

exports[`util/git/author parseGitAuthor parses bot email 1`] = `
Object {
  "address": "some[bot]@users.noreply.github.com",
  "name": "some[bot]",
}
`;

exports[`util/git/author parseGitAuthor parses bot name and email 1`] = `
Object {
  "address": "some[bot]@users.noreply.github.com",
  "name": "some[bot]",
}
`;
